[inductor] add a threshold for membw saving during fusion

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136702
* #138429
* __->__ #136782

Fix https://github.com/pytorch/pytorch/issues/133242 . In that issue, inductor fuses 2 nodes because they access the same scalar tensor. This saving is very small (4 bytes), and if we ignore that, by default, we can not fuse. But if loop ordering after fusion get kicked in, we can reorder loops and fuse those 2 nodes. We get 33% memory bandwidth savings .

I think adding a threshold for membw saving in general is not bad.

I'll run a perf test. ( https://github.com/pytorch/pytorch/actions/runs/11375421752 )



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

shunting314 committed:
[inductor] add a threshold for membw saving during fusion

[ghstack-poisoned]

vkuzo commented on torch/_inductor/config.py:
```diff
@@ -427,6 +427,17 @@ def use_autoheuristic(name: str) -> bool:
+# brings more savings.
+#
+# For the cases loop ordering after fusion does not help, we don't lose much.
+score_fusion_memory_threshold = 10
```
n00b question, is the unit here number of elements, number of bytes, or something else? mostly for my own learning.

shunting314 replied:
It's number of bytes.

eellison reviewed (commented):
Even if the actual memory savings are small, kernel launch overhead can accumulate. I think still makes sense to fuse small tensors

shunting314 commented:
> Even if the actual memory savings are small, kernel launch overhead can accumulate. I think still makes sense to fuse small tensors

It's true. But I think it may not matter much in practice.
1. This scenario can be rare
2. We can leverage CUDAGraphs

Fusing two nodes having very small common memory access looks no much difference to 'fusing two unrelated nodes'. If such fusion in general is good, I guess we should enable 'aggressive_fusion' by default?

shunting314 committed:
Update on "[inductor] add a threshold for membw saving during fusion"


Fix https://github.com/pytorch/pytorch/issues/133242 . In that issue, inductor fuses 2 nodes because they access the same scalar tensor. This saving is very small (4 bytes), and if we ignore that, by default, we can not fuse. But if loop ordering after fusion get kicked in, we can reorder loops and fuse those 2 nodes. We get 33% memory bandwidth savings .

I think adding a threshold for membw saving in general is not bad.

I'll run a perf test. ( https://github.com/pytorch/pytorch/actions/runs/11058587412 )



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

shunting314 committed:
Update

[ghstack-poisoned]

eellison commented:
When running the failing test, `test_asgd_recompile_foreach`, and bumping up the tensor sizes to be more reasonable ([1024, 1024]) this pr makes the test 3x slower. havent debugged further but its possible that not fusing the small fusion is inhibiting a larger fusion. needs more investigation before landing.

shunting314 commented:
> When running the failing test, test_asgd_recompile_foreach, and bumping up the tensor sizes to be more reasonable ([1024, 1024]) this pr makes the test 3x slower. havent debugged further but its possible that not fusing the small fusion is inhibiting a larger fusion. needs more investigation before landing.

In my tests, with the PR the test run time increases from 4.8s to 5.6s. That's probably due to more kernels being generated. But the 'torch._inductor.metrics.num_bytes_accessed' metrics decreased from 117555520 to 83968512  (1.4x reduction) which should be good for perf.

This tests involves some scalar tensors according to the generated wrapper ( https://gist.github.com/shunting314/8421fc964a1fc2848db273cce3d8a5ca ) even if the tensor sizes is increased to be more reasonable [1024, 1024]. I'll force the threshold to be an no-op for those tests and rerun a perf test. Will dig further if the perf test shows some red flag.

shunting314 committed:
Update on "[inductor] add a threshold for membw saving during fusion"


Fix https://github.com/pytorch/pytorch/issues/133242 . In that issue, inductor fuses 2 nodes because they access the same scalar tensor. This saving is very small (4 bytes), and if we ignore that, by default, we can not fuse. But if loop ordering after fusion get kicked in, we can reorder loops and fuse those 2 nodes. We get 33% memory bandwidth savings .

I think adding a threshold for membw saving in general is not bad.

I'll run a perf test. ( https://github.com/pytorch/pytorch/actions/runs/11058587412 )



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented:
Add the [perf result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2010%20Oct%202024%2018%3A33%3A57%20GMT&stopTime=Thu%2C%2017%20Oct%202024%2018%3A33%3A57%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/shunting314/177/head&lCommit=a4ec55bd8d0bd3b67c7f65ffb7144d29bf7951bb&rBranch=main&rCommit=7480e6938d68b7ab51a713aba5c590e8f9fabcb2) here.

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge -i

shunting314 commented:
@pytorchbot merge -i

ezyang commented:
FYI, this is a compile time regression on some HuggingFace models on training:

![image](https://github.com/user-attachments/assets/abf02eaf-67a8-4470-9d74-748d9083b342)

Sounds like it's worth it, but letting you know, maybe there is potentially some compile time optimization opportunities nearby.

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but test_memory starts to fail after this lands in trunk' -c nosignal

inductor/test_memory.py::TestOperatorReorderForPeakMemory::test_reorder_peak_memory [GH job link](https://github.com/pytorch/pytorch/actions/runs/11412837975/job/31760342730) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/6647320de2077c10309f5025a007d51c7fb542d8)

shunting314 commented:
> maybe there is potentially some compile time optimization opportunities nearby.

An alternative to the fix in this PR is to try loop ordering after fusion even if we are able to fuse. But this potentially has larger compilation time hit.

Loop ordering after fusion is off by default now due to compilation time increase. I'll need debug that anyway and hopefully can find some good solutions to improve compilation time.

shunting314 committed:
Update on "[inductor] add a threshold for membw saving during fusion"


Fix https://github.com/pytorch/pytorch/issues/133242 . In that issue, inductor fuses 2 nodes because they access the same scalar tensor. This saving is very small (4 bytes), and if we ignore that, by default, we can not fuse. But if loop ordering after fusion get kicked in, we can reorder loops and fuse those 2 nodes. We get 33% memory bandwidth savings .

I think adding a threshold for membw saving in general is not bad.

I'll run a perf test. ( https://github.com/pytorch/pytorch/actions/runs/11375421752 )



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge -i

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge -i

shunting314 commented:
@pytorchbot merge -f "Try harder.."

ezyang commented:
Final HF compile time regression:

<img width="392" alt="image" src="https://github.com/user-attachments/assets/9a15050d-6d6b-4264-a31a-e63591c9f452">

jansel commented:
This PR makes `x.sum() + 1` become two kernels, because it only saves 4 bytes of memory.  This is likely why it is increasing compile times since we are generating more kernels.

jansel commented:
Possible fix over at #138970
